src/tools/cephfs/top/cephfs-top

```diff
@@ -644,116 +591,209 @@ class FSTop(object):
-            self.stdscr.clear()
-            self.refresh_window_size()
+            if cmd in [ord('m'), ord('q')]:
+                self.set_option(cmd)
+                self.exit_ev.set()
 
+            # header display
             global fs_list
             fs_list = self.get_fs_names()
             stats_json = self.perf_stats_query()
-            self.flag = 0
-            help_text = " [Press 'q' to go back to home (all filesystem info) screen"\
-                        " and 'm' to select another filesystem]"
-
-            # coordinate constants for windowing -- (height, width, y, x)
-            # NOTE: requires initscr() call before accessing COLS, LINES.
-            try:
-                HEADER_WINDOW_COORD = (4, self.width - 1, 0, 0)
-                self.header = curses.newwin(*HEADER_WINDOW_COORD)
-                if fs not in fs_list:
-                    self.stdscr.refresh()
-                    headline = "INFO: Selected filesystem is not available now" + help_text
-                    self.refresh_header(stats_json, headline)
-                else:
-                    if self.height >= 5:
-                        TOPLINE_WINDOW_COORD = (3, self.width - 1, 4, 0)
-                        self.topl = curses.newwin(*TOPLINE_WINDOW_COORD)
-                    else:
-                        self.topl = None
-                    if self.height >= 8:
-                        MAIN_WINDOW_COORD = (self.height - 7, self.width - 1, 7, 0)
-                        self.mainw = curses.newwin(*MAIN_WINDOW_COORD)
-                    else:
-                        self.mainw = None
-                    headline = "HELP: Selected filesystem info" + help_text
-                    if self.refresh_header(stats_json, headline):
-                        x_coord_map = self.refresh_top_line_and_build_coord()
-                        self.refresh_main_window(x_coord_map, stats_json, fs)
-                curses.halfdelay(self.refresh_interval_secs * 10)
-                c = self.header.getch()
-                if c in [ord('m'), ord('q')]:
-                    self.set_option(c)
-            except curses.error:
-                # this may happen when addstr the terminal window
-                # size changed, just retry it
-                pass
+            if fs not in fs_list:
+                help = "Error: The selected filesystem is not available now. " + help_commands
+                self.header.erase()  # erase previous text
+                self.refresh_header(stats_json, help, screen_title, 3)
+            else:
+                self.tablehead_y = 0
+                help = "HELP: " + help_commands
+                self.fsstats.erase()  # erase previous text
+
+                if self.refresh_header(stats_json, help, screen_title, 3):
+                    x_coord_map = self.refresh_top_line_and_build_coord(self.tablehead_y)
+                    self.refresh_clients(x_coord_map, stats_json, fs)
+
+            # scroll and refresh
```

Comment:
Currently, we are having a hardcoded value of `self.PAD_HEIGHT` due to which on vertical scrolling, we keep going until we reach `self.PAD_HEIGHT` value, even when we have very less metrics to display. We can add something like `self.PAD_HEIGHT = self.tablehead_y + self.header_height` before `line:660` to achieve a self adjusting `self.PAD_HEIGHT` value which will allow us to scroll till our metrics display ends.Similarly, something like this for `run_display`.I can see, its mentioned in desc. that we configure it later, but I don't think we have something to wait for this.It will make the scroll feature better.

Reply:
I don't think adjusting the pad height every time is a good idea to have in a user interface. It could be configured, but not adjustable based on the items to be displayed which is what we were doing in the old cephfs-top.

That said, this issue is already fixed in the latest code. You should checkout the latest code.

Reply:
I checked the latest code, the above mentioned issue is fixed.But, there are cases when we aren't able to see all the filesystem metrics as scroll doesn't allow to move below a specific window size.

Reply:
> I don't think adjusting the pad height every time is a good idea to have in a user interface. It could be configured, but not adjustable based on the items to be displayed which is what we were doing in the old cephfs-top.
> 
> That said, this issue is already fixed in the latest code. You should checkout the latest code.

We can handle the pad height separately inside both display functions for the comparison in `KEY_UP` and `KEY_DOWN` instead of using the `self.PAD_HEIGHT`, with reference to the way it was done before latest change.

Reply:
> > I don't think adjusting the pad height every time is a good idea to have in a user interface. It could be configured, but not adjustable based on the items to be displayed which is what we were doing in the old cephfs-top.
> > That said, this issue is already fixed in the latest code. You should checkout the latest code.
> 
> We can handle the pad height separately inside both display functions for the comparison in `KEY_UP` and `KEY_DOWN` instead of using the `self.PAD_HEIGHT`, with reference to the way it was done before latest change.

@neesingh-rh 
I request you to understand what's the need for a pad and how it differs from a window. You should try to understand why a pad height is not meant to adjust dynamically and create a new pad every time we get new metrics. Scrolling is not about adjusting the pad height or pad width. Then what is it about? Please understand that part.

> But, there are cases when we aren't able to see all the filesystem metrics as scroll doesn't allow to move below , at a specific window size.

Well, the user should have a sensible screen size first, before asking why it's not showing the last few rows. I meant, what's the use case of having such a tiny size window? And what prevents the user to drag and increase the window size a little bit? In real-time, the user would have larger window sizes instead. I've checked out this particular issue and this happens only when the window size is tiny, in which curses doesn't have enough room to make a scrolling-end action. In fact, it does it internally. But the user doesn't allow himself to see it.

Reply:
> > But, there are cases when we aren't able to see all the filesystem metrics as scroll doesn't allow to move below , at a specific window size.
> 
> Well, the user should have a sensible screen size first, before asking why it's not showing the last few rows. I meant, what's the use case of having such a tiny size window? And what prevents the user to drag and increase the window size a little bit? In real-time, the user would have larger window sizes instead. I've checked out this particular issue and this happens only when the window size is tiny, in which curses doesn't have enough room to make a scrolling-end action. In fact, it does it internally. But the user doesn't allow himself to see it.

Yeah, Make sense. I checked again, this issue comes into picture only when there is a tiny window size.Otherwise, its working fine.

Reply:
It works with tiny sized window too. But if the user doesn't give enough viewport size to see the scroll-end action, he can't see it ends. There's no shrink and show mechanism involved and we can't use a different offset increment value as the viewport shrinks.